### ISSUE TITLE
PYTHONPATH not set for dependencies in a `BundlePackage`

### ISSUE THREAD

**Issue by shuds13** [LLNL/spack#32238]
### Steps to reproduce

If I install a Python package in a Spack environment the view is added to PYTHONPATH. If I install xsdk (which has Python dependencies), it is not. Also, the installed python is not on my PATH.

```console

spack env create myenv
spack env activate -p myenv
spack add xsdk
spack install
echo $PYTHONPATH

# Try reactivating
despacktivate
spack env activate -p myenv
echo $PYTHONPATH

```

For faster test, you can trim down the xsdk package.py so just uses py-libensemble (this is the python package).

e.g. you can trim main function to 

```python
class Xsdk(BundlePackage,):
    version("0.7.0")
    variant("libensemble", default=True, description="Enable py-libensemble package build")
    depends_on("py-libensemble")
```

I've tried activating with -v also. I also tried adding type=("run") to the depends_on line.

### Error message

It cannot find python packages unless I do a `spack load`.

### Information on your system

Ubuntu 20.04

[myenvt] shudson@shuds:test_spack_env$ spack debug report
* **Spack:** 0.19.0.dev0 (8f34a3ac5c7ff09751007d57c88884aee48921bb)
* **Python:** 3.9.5
* **Platform:** linux-ubuntu20.04-skylake
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by shuds13** [comment]
Update - if I change the above class line to also inherit from PythonPackage, it looks like PYTHONPATH does get set. 

```python
class Xsdk(BundlePackage, PythonPackage):
    version("0.7.0")
    variant("libensemble", default=True, description="Enable py-libensemble package build")
    depends_on("py-libensemble", type=("run"))
```

So I guess the question is what is the desired bahavior. I would think, you would always want python dependencies to be added to PYTHONPATH.

**Comment by adamjstewart** [comment]
`PYTHONPATH` isn't controlled by `PythonPackage`, it's controlled by `extends('python')`. If packages use `extends('python')`, regardless of whether or not they inherit from `PythonPackage`, they should be added to `PYTHONPATH`. Note that there is a bug wherein packages that extend multiple libraries (Python, R, Perl, etc.) may not be added to `PYTHONPATH`, but as far as I know that should only affect a few packages.

**Comment by shuds13** [comment]
@adamjstewart 

I can confirm adding extends('python') worked in both my simplified example above, and in full xsdk. The actual xsdk package does in fact use multiple inheritence `class Xsdk(BundlePackage, CudaPackage, ROCmPackage):`, but it still seems to work. 

Note, I do find I have to despacktivate and reactivate the environment after reconcretizing for it to work.

**Comment by shuds13** [comment]
Just to add a thought, in my naive view, it would be good if any package with python dependencies (at least run dependencies) updated the PYTHONPATH. I don't know all the practical considerations with that.

**Comment by adamjstewart** [comment]
`extends('python')` is identical to `depends_on('python')` but also sets `PYTHONPATH`. There are thousands of packages that depend on Python but don't install any Python libraries.

**Comment by balay** [comment]
@adamjstewart xsdk is just a metapackage. py-libensemble is the python package here. As @shuds13 indicated xsdk just adds a dependency 

```python
class Xsdk(BundlePackage, PythonPackage):
    depends_on("py-libensemble", type=("run"))
```

So I would think that `extends('python')` should go into `py-liensemble` and not xsdk. Is this not the case? [i.e spack env requires this in xsdk - not py-liensemble?]

**Comment by adamjstewart** [comment]
Correct, and `py-libensemble` is a subclass of `PythonPackage`, so it already has `extends('python')`. I don't know why a metapackage would behave differently, but this should work as is.

**Comment by shuds13** [comment]
Yes, it doesn't work for me, I have to add the extends('python') to the xsdk package. My steps taken are at the top, if anyone wants to try.

**Comment by adamjstewart** [comment]
@becker33 @scheibelp any thoughts on this? It seems like `PYTHONPATH` isn't being set for dependencies in an environment.
